### DISCUSSION TITLE
Use PEP 604 union types syntax instead of `Optional` as `Mapped` parameter?

### DISCUSSION THREAD

**Discussion by homeworkprod** [sqlalchemy/sqlalchemy#10525] [Usage Questions] [answered]
I can't seem to find this mentioned anywhere in the docs, discussions, StackOverflow, or elsewhere, so I'm asking here:

Is it supported to use the [PEP 604](https://peps.python.org/pep-0604/) union types syntax with `Mapped` in SQLAlchemy 2.x, i.e. instead of 
```py
description: Mapped[Optional[str]] = mapped_column(  # […]
```
write this instead?
```py
description: Mapped[str | None] = mapped_column(  # […]
```

Are there (unapparent) limitations why this won't work? Or is it just to use syntax in the documentation that works on Python before 3.10?

**Comment by zzzeek** [comment, marked as answer]
all union types should be interpreted in the same way as Optional with no difference in behavior.

**Reply by homeworkprod** [reply]
Thanks for the answer, Mike!

I can confirm that using union types syntax works well (for me).

**Comment by AbdealiLoKo** [comment]
When I try this, I seem to be getting this error:
```
sqlalchemy.exc.ArgumentError: Could not resolve all types within mapped annotation: "Mapped[str | None]".  Ensure all types are written correctly and are imported within the module in use.
```

I am using:
- sqlalchemy 2.0.27
- python 3.7.16

Could this be because I use py3.7 ?

PS: I always use `from __future__ import annotations` ... So that I can use the latest typing syntax and pylance can parse that for me.

**Reply by zzzeek** [reply]
it's py3.7.  We run these annotations through real evaluate at runtime so the pipe character will fail on Py3.7.

**Reply by AbdealiLoKo** [reply]
That makes sense -_-
I realized that about 5 mins after I posted --> And looks like ruff already [solves for this](https://docs.astral.sh/ruff/settings/#lint_pyupgrade_keep-runtime-typing) 

Sorry about the noise !
